### PR TITLE
iox-#734 Update hoofs readme for v2.0 release

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

Updates the iceoryx_hoofs readme

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Relates to #743

### CODE REVIEW

File: iceoryx_hoofs/README.md [hunk 1]
```diff
@@ -1,68 +1,76 @@
...
 
-STL constructs which are not in the C++11 standard included are contained here as
-well as constructs which are helping us in our daily life. We differ in
-here from our coding guidelines and follow the C++ STL coding guidelines
-to help the user to have a painless transition from the official STL
-types to ours. The API should also be identical to the corresponding
-STL types but we have to make exceptions here. For instance, we do not
-throw exceptions, try to avoid undefined behavior and we do not use
-dynamic memory. In these cases we adjusted the API to our use case.
+This contains STL constructs not part of the C++14 standard as well as convenience
```
Comment (elfenpiff) [inline iceoryx_hoofs/README.md:31]: STL constructs which are not part of the C++14

File: iceoryx_hoofs/README.md [hunk 2]
```diff
@@ -1,68 +1,76 @@
...
+|`helplets`           |   |   | Implementations of [C++ Core Guideline](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines) concepts like `not_null` are contained here. Additionally, we are providing some types to verify preconditions at compile time. Think of an int which has to be always greater 5, here we provide types like `greater_or_equal<int, 6>`.|
 |`list`               |   |   | Heap and exception free, relocatable implementation of `std::list` |
-|`MethodCallback  `  |   |   | Constructs a MethodCallback from a pointer to a specific object and a pointer to a method of that object, also as `ConstMethodCallback` available |
+|`MethodCallback`     |   |   | Constructs a callback from a pointer to a specific object and a pointer to a method of that object, also as `ConstMethodCallback` available |
```
Comment (elfenpiff) [inline iceoryx_hoofs/README.md:57]: Could you please add a check for maybe obsolete. With the `cxx::function` at our hands and capturing we do not need this anymore.

File: iceoryx_hoofs/README.md [hunk 3]
```diff
@@ -155,28 +167,28 @@ setTimeout(11_s); // 11 seconds
...
+|:---------------------:|:--------:|:--------------:|:------------|
+|`ObjectPool`           | i |   | Container which stores raw objects without calling the ctor of the objects.  |
 
 ### graphs
 
 | class                   | internal | maybe obsolete | description |
 |:-----------------------:|:--------:|:--------------:|:------------|
-|`DirectedGraph`          | i | | Creates and manages a [directed graph](https://en.wikipedia.org/wiki/Directed_graph).  |
-|`DirectedAcyclicGraph`   | i | | Like the DirectedGraph but additional checks prohibit to create edges which have a cyclic dependency.  |
+|`DirectedGraph`          | i |   | Creates and manages a [directed graph](https://en.wikipedia.org/wiki/Directed_graph).  |
```
Comment (elfenpiff) [inline iceoryx_hoofs/README.md:184]: Where do we use this? Would it make sense to make this obsolete?

File: iceoryx_hoofs/README.md [hunk 4]
```diff
@@ -1,68 +1,76 @@
...
-| [design_pattern](#design-pattern) | Certain code patterns which are repeating themselves all over the code are abstracted and stored in here. At the moment we only have the creation pattern but maybe there will be more. |
-| [error-handling](#error-handling) | The Error-Handler to in iceoryx for handling fault cases. |
+| [concurrent](#concurrent) | You should never use concurrent constructs like `mutex`, `semaphores`, `atomic` etc directly in our code-base. At the moment we still have some exceptions to this guideline but the idea is that all classes which are using them are stored under concurrent and have to undergo more tests then the usual non concurrent class. For instance we try to provide stress tests for them. This module provides classes like `fifo`, `smart_lock`, `sofi`, `trigger_queue` and much more. |
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/README.md:13]: But my spellchecker insisted on `code-base` :)

File: iceoryx_hoofs/README.md (hunk 3 above)
Comment (mossmaurice) [reply iceoryx_hoofs/README.md:186]: So far I had in mind to wrap all "maybe obsolete" building blocks their own namespace, to be able to detect via `objdump` if some of those classes have been added to the final binary via `tools/scripts/analyse_blacklisted_namespaces_in_binary.sh`. Any objections to this idea?

File: iceoryx_hoofs/README.md [hunk 5]
```diff
@@ -1,68 +1,76 @@
...
-| [design_pattern](#design-pattern) | Certain code patterns which are repeating themselves all over the code are abstracted and stored in here. At the moment we only have the creation pattern but maybe there will be more. |
-| [error-handling](#error-handling) | The Error-Handler to in iceoryx for handling fault cases. |
+| [concurrent](#concurrent) | You should never use concurrent constructs like `mutex`, `semaphores`, `atomic`, etc directly in our codebase. At the moment we still have some exceptions to this guideline but the idea is that all classes which are using them are stored under concurrent and have to undergo more tests then the usual non concurrent class. For instance we try to provide stress tests for them. This module provides classes like `fifo`, `smart_lock`, `sofi`, `trigger_queue` and much more. |
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/README.md:13]:
```suggestion
| [concurrent](#concurrent) | You should never use concurrent constructs like `mutex`, `semaphores`, `atomic`, etc. directly in our codebase. At the moment we still have some exceptions to this guideline but the idea is that all classes which are using them are stored under concurrent and have to undergo more tests then the usual non concurrent class. For instance we try to provide stress tests for them. This module provides classes like `fifo`, `smart_lock`, `sofi`, `trigger_queue` and much more. |
```

File: iceoryx_hoofs/README.md [hunk 6]
```diff
@@ -93,60 +103,64 @@ attribute overview of the available Queues:
...
-|`Command`           | i | | Base class for void(Arg) calls - known as commands. |
+| class               | internal | maybe obsolete | description |
+|:-------------------:|:--------:|:--------------:|:------------|
+|`Creation`           |   | X | When implementing resource handling classes which follow the RAII idiom we may have to throw exceptions inside the constructor. As an alternative to exceptions we have the creation pattern, a specialized factory which returns the object inside of an `expected`. |
 
 ### Error handling
 
-The Error-Handler is a central instance for collecting al errors and react to them. In the file `error-handling.hpp` are all error enums collected. The Error-Handler has different error-levels, for more information see [error-handling.md](https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/design/error-handling.md)
+The error handler is a central instance for collecting all errors and react to them. The `error-handling.hpp` contains a list of all error enum values. The error handler has different error-levels, for more information see [error-handling.md](https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/design/error-handling.md)
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/README.md:118]:
```suggestion
The error handler is a central instance for collecting all errors and react to them. The `error-handling.hpp` contains a list of all error enum values. The error handler has different error levels, for more information see [error-handling.md](https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/design/error-handling.md)
```

File: iceoryx_hoofs/README.md (hunk 5 above)
Comment (elBoberido) [reply iceoryx_hoofs/README.md:13]: Oh boy, I need some vacation. While my eyes moved from your suggestion to the code I forgot about the period and added a semicolon before `etc`

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by elfenpiff — APPROVED** [review]

**Pull request merged**